enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,782 @@
+1. The Cluster API controller takes over management of the MachineSet going forwards
+
+To migrate back from Cluster API to Machine API, the procedure is the same, however, the value of the field should be set to `machine.openshift.io`.
+
+When the conversion cannot proceed for some reason, for example, a feature is in use in Machine API that is not present in Cluster API, the following will happen:
+1. Cluster admin identifies a MachineSet to migrate to Cluster API
+1. Cluster admin uses `oc edit` to update the value of the `spec.authoritativeAPI` field to `cluster.x-k8s.io`
+1. The sync controller verifies that the move from Machine API to Cluster API is not valid and that the Cluster API version of the resource is not fully synced with the Machine API version
+1. The sync controller adds a condition to the MachineSet to indicate that the migration cannot proceed
+1. The `status.authoritativeAPI` field is not updated and the `status.synchronizedGeneration` is not updated
+1. Where appropriate, the cluster admin contacts support for additional help/timelines for availability of the missing feature
+1. The Machine API controllers continue to manage the MachineSet
+
+#### Workflow extension
+
+When migration between the two authoritative API versions is not possible, the sync controller will add a condition to the resource to indicate the reason for the failure.
+However, with the above described workflow, the user is not presented with a synchronous response to the migration request.
+
+To improve the feedback loop, a webhook will be introduced to validate the migration request.
+
+The migration request will always be persisted, however, the webhook will check for the presence of the sync controller's failure condition, and return a warning response to the user if the migration is not possible.
+
+The users intention to migrate will be persisted, and the sync controller will continue to attempt to migrate the resource until the condition is resolved, but, the user is also informed of the reason for the failure synchronously.
+
+This extension is not required for the minimum viable version of this project, but, will be added in a future iteration, prior to the GA release.
+
+### API Extensions
+
+This enhancement introduces a new `authoritativeAPI` field on `spec` and `status` of Machine API resources.
+Users will set the value of this field to determine which of the Machine API or Cluster API controllers will own the resource.
+
+The `status` will also include a `synchronizedGeneration` field to indicate the generation of the authoritative resource that the non-authoritative resource is synchronised with.
+
+Machine API conditions will also include a new `Synchronized` condition type, which will be used to indicate when the non-authoritative resource is fully synchronised with the authoritative resource.
+
+#### Authority of resources and controller reconciliations
```

Comment:
I expected to see discussion of an admission plugin here that rejects updates to non-authoritative except for
1. non-functional label and annotation updates
2. sync controller

You'll really want clients to told very clearly that their changes aren't working at the point where they fail.

Reply:
This would also give a way to provide a metrics comparison of before and after for rejection to have some confidence that writers aren't broken.

Reply:
Yes, I think that's a reasonable ask. I'll ad that in

Reply:
Added a section on admission